src/interpreter/expression-iterator.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2025 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef interpreter_expression_iterator_h
+#define interpreter_expression_iterator_h
+
+#include <cassert>
+#include <iterator>
+#include <vector>
+
+#include "wasm.h"
+
+namespace wasm::interpreter {
+
+// TODO: This is a quick and dirty hack. We should implement a proper iterator
+// in ir/iteration.h that keeps only a vector of (Expression*, index) pairs to
+// find the current location in the epxression tree.
+struct ExpressionIterator {
+  using difference_type = std::ptrdiff_t;
+  using value_type = Expression*;
+  using pointer = Expression**;
+  using reference = Expression*&;
+  using iterator_category = std::input_iterator_tag;
+
+  // The list of remaining instructions in reverse order so we can pop from the
+  // back to advance the iterator.
+  std::vector<Expression*> exprs;
```

Comment:
How will this work with loops etc.?

Generally, I was guessing this would use BinaryenIR itself. We could track the location of the frame pointer using a simple indexing, a vector of indexes basically,
```
[ index at toplevel, index at the higher level, index at a higher level still, ...]
```
So `[1, 0]` would mean we are at the second instruction at the toplevel, and the first of its children, etc. If we keep pointers each of the children then we can quickly traverse between them etc.

Reply:
Oh, actually we have ChildIterator already. So a stack of ChildIterators could work?

Reply:
Yep, that's what the TODO above the class is about. If for some reason we wanted to implement loops before implementing that better expression iterator, we would have to copy the vector-based iterator at each loop header and store it in a map on the frame. There are TODOs for those parts in `Frame` and the `ExpressionIterator` constructor. It's probably better to implement the better iterators first, though.

All other control flow transfers are forward edges, so they are very simple with the vector-based iterators. You just increment the iterator until you find the block matching the target.

Reply:
Yes, a stack of ChildIterators would work if we don't mind the memory overhead of materializing the children of each expression in the current path. Alternatively we could do something more specialized that only stores an `Expression*` and an `Index` for each expression in the path.

Reply:
What do you mean by "materializing" here? I mean that I'm not sure what overhead ChildIterator has that this does not.

Reply:
`ChildIterator` contains `SmallVector<Expression**, 4> children;`. In principle it could be reimplemented to contain only a single `Expression*` and an index.

Reply:
But then each step would need to find the children again? That seems likely to be slower.

I might think otherwise if memory caching issues were possible, like it we were storing large amounts of such vectors. But typically they would be quite small?

Reply:
Though, I am open to being convinced otherwise of course! And if that is faster, we can implement that in ChildIterator as an optimization?

Reply:
Yeah, it might not matter, and if it does, improving ChildIterator would make sense.

Reply:
I've updated the TODO to capture all of this.